### ISSUE TITLE
Add benchmarks and generated ASM code comparison in the docs

### ISSUE THREAD

**Issue by Veetaha** [elastio/bon#4]
# Context

This is from [the Reddit post](https://www.reddit.com/r/rust/comments/1eeem92/comment/lfe1ezr/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button).

Developers that would like to use `bon` in performance-sensitive code would like to have some benchmarks and proofs of the low cost of using this abstraction. This should make it easy for them to rely on `bon`.

# Summary

We need to have some reasonable benchmark showcase and ASM code comparison in the docs. It should show the difference between calling a function with positional parameters and using the `bon::builder`. Ideally `rustc` should optimize all of this and there should be zero difference, but it needs to be experimentally proven

**Comment by LockedThread** [comment]
I just wanted to say, awesome job making this library. I am looking forward to these benchmarks.

**Comment by Veetaha** [comment]
I've done some ASM comparison in godbolt. I pasted the vanilla positional syntax of function call and the code that `bon::builder` macro generates. I manually expanded the macro and pasted the generated code into godbolt with small 
nonessential changes to make it compile there.

This is just my first dirty try before I test several more advanced cases. However, with `opt-level=3` there is literally no differences between the positional function and builder versions. The compiler did a good job of inlining the functions and removing the builder overhead entirely.

Here is the link to [godbolt](https://godbolt.org/z/dT89Tncbr)

![image](https://github.com/user-attachments/assets/9423c7b4-6f87-4322-8f7d-923744db1ace)
